File: puzzles/ts/shadows-knight1/shadows-knight1.ts
let inputs: string[] = readline().split(' ');
const width: number = Number(inputs[0]); // width of the building.
const height: number = Number(inputs[1]); // height of the building.
const n: number = Number(readline()); // maximum number of turns before game over.
inputs = readline().split(' ');
let x: number = Number(inputs[0]);
let y: number = Number(inputs[1]);

let minBombX = 0;
let maxBombX = width - 1;
let minBombY = 0;
let maxBombY = height - 1;

// game loop
while (true) {
  const direction = readline(); // the direction of the bombs from batman"s current location (U, UR, R, DR, D, DL, L or UL)

  switch (direction) {
    case "U":
      minBombX = maxBombX = x;
      maxBombY = y - 1;
      break;
    case "R":
      minBombY = maxBombY = y;
      minBombX = x + 1;
      break;
    case "D":
      minBombX = maxBombX = x;
      minBombY = y + 1;
      break;
    case "L":
      minBombY = maxBombY = y;
      maxBombX = x - 1;
      break;
    case "UR":
      minBombX = x + 1;
      maxBombY = y - 1;
      break;
    case "UL":
      maxBombX = x - 1;
      maxBombY = y - 1;
      break;
    case "DR":
      minBombX = x + 1;
      minBombY = y + 1;
      break;
    case "DL":
      maxBombX = x - 1;
      minBombY = y + 1;
      break;
  }

  x = Math.floor((minBombX + maxBombX) / 2);
  y = Math.floor((minBombY + maxBombY) / 2);

  // the location of the next window Batman should jump to.
  console.log(`${x} ${y}`);
}

File: puzzles/kotlin/src/shadows-knight1/shadows-knight1.kt
import java.util.*

fun main(args : Array<String>) {
  val input = Scanner(System.`in`)
  val width = input.nextInt() // width of the building.
  val height = input.nextInt() // height of the building.
  val n = input.nextInt() // maximum number of turns before game over.
  var x = input.nextInt()
  var y = input.nextInt()

  var minX = 0
  var maxX = width -1
  var minY = 0
  var maxY = height - 1

  // game loop
  while (true) {
    val bombDir = input.next() // the direction of the bombs from batman's current location

    when (bombDir) {
      "U" -> {
        maxY = y - 1
        minX = x
        maxX = x
      }

      "UR" -> {
        minX = x + 1
        maxY = y - 1
      }

      "R" -> {
        minX = x + 1
        minY = y
        maxY = y
      }

      "DR" -> {
        minX = x + 1
        minY = y + 1
      }

      "D" -> {
        minY = y + 1
        minX = x
        maxX = x
      }

      "DL" -> {
        maxX = x - 1
        minY = y + 1
      }

      "L" -> {
        maxX = x - 1
        minY = y
        maxY = y
      }

      // UL
      else -> {
        maxX = x - 1
        maxY = y - 1
      }
    }

    x = (minX + maxX) / 2
    y = (minY + maxY) / 2
    // the location of the next window Batman should jump to.
    println("$x $y")
  }
}
